File: packages/core/src/llm/claude-client.ts
import Anthropic from '@anthropic-ai/sdk';
import { createLogger } from '../logging/logger.js';
import { TokenBudgetError, RateLimitError, AuthError, NetworkError } from '../errors.js';
import { parseJSONResponse } from './json-extract.js';

const log = createLogger('ClaudeClient');

export interface ClaudeResponse {
  content: string;
  usage: { inputTokens: number; outputTokens: number };
}

/**
 * Claude API 클라이언트 인터페이스. 테스트에서 mock 주입 가능.
 */
export interface IClaudeClient {
  /** 현재까지 사용한 총 토큰 수 */
  readonly tokensUsed: number;

  chatJSON<T>(
    systemPrompt: string,
    userMessage: string,
  ): Promise<{
    data: T;
    usage: ClaudeResponse['usage'];
  }>;

  chat(
    systemPrompt: string,
    userMessage: string,
  ): Promise<ClaudeResponse>;
}

export interface ClaudeClientConfig {
  model: string;
  maxTokens: number;
  temperature: number;
  maxRetries?: number;
  /** Token budget. 0 = unlimited. */
  tokenBudget?: number;
}

const DEFAULT_MAX_RETRIES = 3;
const BASE_DELAY_MS = 1000;
const JITTER_FACTOR = 0.2; // 0~20% 랜덤 jitter

/**
 * Anthropic Claude API 공유 클라이언트.
 * Worker Agent들이 공통으로 사용한다. Token budget 추적 포함.
 */
export class ClaudeClient implements IClaudeClient {
  private client: Anthropic;
  private config: ClaudeClientConfig;
  private totalTokensUsed = 0;

  constructor(config: ClaudeClientConfig, apiKey: string) {
    this.client = new Anthropic({ apiKey });
    this.config = config;
  }

  /** 현재까지 사용한 총 토큰 수 */
  get tokensUsed(): number {
    return this.totalTokensUsed;
  }

  async chat(systemPrompt: string, userMessage: string): Promise<ClaudeResponse> {
    // Token budget 체크
    const budget = this.config.tokenBudget ?? 0;
    if (budget > 0 && this.totalTokensUsed >= budget) {
      throw new TokenBudgetError(this.totalTokensUsed, budget);
    }

    const response = await this.withRetry(async () => {
      const result = await this.client.messages.create({
        model: this.config.model,
        max_tokens: this.config.maxTokens,
        temperature: this.config.temperature,
        system: systemPrompt,
        messages: [{ role: 'user', content: userMessage }],
      });

      const textBlock = result.content.find((b) => b.type === 'text');
      return {
        content: textBlock ? textBlock.text : '',
        usage: { inputTokens: result.usage.input_tokens, outputTokens: result.usage.output_tokens },
      };
    });

    // Token 사용량 누적
    this.totalTokensUsed += response.usage.inputTokens + response.usage.outputTokens;
    log.info(
      { inputTokens: response.usage.inputTokens, outputTokens: response.usage.outputTokens, totalUsed: this.totalTokensUsed },
      'Claude usage',
    );

    return response;
  }

  async chatJSON<T>(
    systemPrompt: string,
    userMessage: string,
  ): Promise<{ data: T; usage: ClaudeResponse['usage'] }> {
    const response = await this.chat(
      systemPrompt + '\n\nIMPORTANT: Respond with valid JSON only. No markdown, no explanation.',
      userMessage,
    );

    const data = parseJSONResponse<T>(response.content, 'Claude');
    return { data, usage: response.usage };
  }

  private async withRetry<T>(fn: () => Promise<T>): Promise<T> {
    const maxRetries = this.config.maxRetries ?? DEFAULT_MAX_RETRIES;
    let lastError: Error | null = null;

    for (let attempt = 0; attempt <= maxRetries; attempt++) {
      try {
        return await fn();
      } catch (error) {
        lastError = error instanceof Error ? error : new Error(String(error));
        if (!this.isRetryable(lastError)) throw lastError;
        if (attempt < maxRetries) {
          const jitter = 1 + Math.random() * JITTER_FACTOR;
          const delay = BASE_DELAY_MS * Math.pow(2, attempt) * jitter;
          log.warn(
            {
              attempt: attempt + 1,
              maxRetries,
              delayMs: Math.round(delay),
              err: lastError.message,
            },
            'Retrying',
          );
          await new Promise((r) => setTimeout(r, delay));
        }
      }
    }

    throw lastError!;
  }

  private isRetryable(error: Error): boolean {
    // 커스텀 에러 타입 우선 체크
    if (error instanceof RateLimitError || error instanceof NetworkError) return true;
    if (error instanceof AuthError || error instanceof TokenBudgetError) return false;

    // Anthropic SDK 에러 또는 알 수 없는 에러 → 메시지 기반 판별
    const msg = error.message.toLowerCase();
    if (msg.includes('rate limit') || msg.includes('429')) return true;
    if (msg.includes('timeout') || msg.includes('timed out')) return true;
    if (msg.includes('network') || msg.includes('econnreset') || msg.includes('socket'))
      return true;
    // 단어 경계(\b)로 숫자가 다른 문자열의 일부일 때 false positive 방지
    if (/\b5\d{2}\b/.test(msg)) return true;
    if (msg.includes('401') || msg.includes('403') || msg.includes('invalid')) return false;
    return false; // unknown errors → do not retry (안전한 기본값)
  }
}

File: backend/src/core/llm/claude_client.py
"""Anthropic API 기반 Claude 클라이언트."""
from __future__ import annotations

import anthropic

from src.core.config import DEFAULT_CLAUDE_MODEL
from src.core.errors import RateLimitError, TokenBudgetError
from src.core.llm.json_extract import parse_json_response
from src.core.logging.logger import get_logger
from src.core.resilience.api_retry import with_retry

log = get_logger("ClaudeClient")

# Claude 모델 화이트리스트 (injection 방지)
ALLOWED_MODELS = {
    "claude-opus-4-20250514",
    "claude-sonnet-4-20250514",
    "claude-sonnet-4-5-20251001",
    "claude-haiku-4-5-20251001",
    "claude-3-5-sonnet-20241022",
    "claude-3-5-haiku-20241022",
    "claude-3-opus-20240229",
}


class ClaudeClient:
    def __init__(self, api_key: str, default_model: str = DEFAULT_CLAUDE_MODEL) -> None:
        self._client = anthropic.AsyncAnthropic(api_key=api_key)
        self._default_model = default_model
        self._tokens_used = 0

    async def chat(
        self,
        messages: list[dict],
        system: str | None = None,
        model: str | None = None,
        max_tokens: int = 4096,
        temperature: float = 0.7,
        token_budget: int | None = None,
    ) -> tuple[str, int, int]:
        """
        Claude API 호출.
        Returns: (response_text, input_tokens, output_tokens)
        """
        resolved_model = model or self._default_model
        if resolved_model not in ALLOWED_MODELS:
            log.warn("Unknown model requested, falling back to default", model=resolved_model)
            resolved_model = self._default_model

        if token_budget and self._tokens_used >= token_budget:
            raise TokenBudgetError(self._tokens_used, token_budget)

        async def _call():
            kwargs = {
                "model": resolved_model,
                "max_tokens": max_tokens,
                "messages": messages,
            }
            if system:
                kwargs["system"] = system
            if temperature != 1.0:
                kwargs["temperature"] = temperature
            return await self._client.messages.create(**kwargs)

        try:
            response = await with_retry(
                _call,
                max_retries=3,
                base_delay_ms=1000,
                label=f"Claude {resolved_model}",
            )
        except anthropic.RateLimitError as e:
            raise RateLimitError("Claude API", cause=e) from e

        input_tokens = response.usage.input_tokens
        output_tokens = response.usage.output_tokens
        self._tokens_used += input_tokens + output_tokens

        text = response.content[0].text if response.content else ""
        return text, input_tokens, output_tokens

    async def chat_json(
        self,
        messages: list[dict],
        system: str | None = None,
        model: str | None = None,
        max_tokens: int = 4096,
        temperature: float = 0.7,
        token_budget: int | None = None,
    ) -> tuple[dict | list, int, int]:
        """JSON 응답을 파싱해서 반환."""
        text, input_tokens, output_tokens = await self.chat(
            messages=messages,
            system=system,
            model=model,
            max_tokens=max_tokens,
            temperature=temperature,
            token_budget=token_budget,
        )
        return parse_json_response(text), input_tokens, output_tokens

    @property
    def tokens_used(self) -> int:
        return self._tokens_used
